### PR TITLE
VFB-81: Concurrent record editing

### PR DESCRIPTION
**Pull request by minalui**
## What's changed
- Add last_updated column to parcels and clients table, default set to current utc
- Existing rows will have last_update set to utc when migration happens
- Add function to updated last_updated to current utc when an update is made to the row
- Change update functions to check last_updated in client side is the same as database
- Handle and display errors

## Screenshots / Videos
| Before     | After      |
|------------|------------|
| paste_here | paste_here |

## Checklist
- [x] The ticket is up-to-date - Please document any deviations from the original approach if there is any.
- [x] I have documented the testing steps for QA
- [x] I have self-reviewed this PR
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [x] Make sure you've fixed all linting problems with `npm run lint_fix`
- [x] Make sure you've tested via `npm run test:e2e`

If you have made any changes to the database...
  - [x] The migration files are up-to-date with my final set up (`npx supabase db diff -f <name_of_migration>` should create nothing at this point)
  - [x] I have updated the typescript definitions for the database with `db:local:generate_types`
  - [ ] I have modified the seed in `supabase/seed/seed.mts` if appropriate N/A
  - [ ] If I have modified the seed, I have also generated the seed with `npm run db:generate_seed`  N/A
  - [ ] With my final set up, I can run `npm run dev:reset_supabase` without any errors. N/A
  - Does this require resetting the deployed database? - NO (remove as appropriate)
  - [x] I have checked that migration can happen successfully without resetting the database, doing the following.
    - Make sure you have rebased your branch onto dev or merged dev into your branch.
    - Checkout the dev branch
    - Run `npm run post_checkout` to reset the database, including the seed data.
    - Checkout your branch
    - Run `npx supabase migrations up --include-all --local` to run all outstanding migration files
    - Check that the resulting database is what you expect, bar any seed data changes.

### CODE REVIEW

File: src/app/clients/form/submitFormHelpers.ts [hunk 1]
```diff
@@ -60,13 +58,19 @@ const formatClientRecord = (
...
+        last_updated: fields.lastUpdated,
     };
 };
 
-export const submitAddClientForm = async (
-    fields: ClientFields,
-    router: AppRouterInstance
-): Promise<void> => {
+type addClientErrors = "failedToInsertClientAndFamily";
+type addClientReturnType =
```
Comment (rickaw) [inline src/app/clients/form/submitFormHelpers.ts:66]: For convention, we'll stick to `...Result` :)
```suggestion
type addClientResult =
```

File: src/app/clients/form/submitFormHelpers.ts [hunk 2]
```diff
@@ -99,21 +103,32 @@ export const submitAddClientForm = async (
         clientId: clientId,
     });
 
-    router.push(`/parcels/add/${clientId}`);
+    return { clientId: clientId, error: null };
 };
 
+type editClientErrors = "failedToUpdateClientAndFamily" | "concurrentUpdateConflict";
+type editClientReturnType =
```
Comment (rickaw) [inline src/app/clients/form/submitFormHelpers.ts:110]:
```suggestion
type editClientResult =
```

**Review by rickaw — CHANGES_REQUESTED** [review]

File: supabase/migrations/20240423141258_update_client_function_change.sql
[the remark this answers is not in the document]
Comment (minalui) [reply supabase/migrations/20240423141258_update_client_function_change.sql:1]: The function now only updates a row if primary_key and last_updated matches. It also returns the client ID and number of rows updated in a new database type.

File: src/app/error.tsx
[the remark this answers is not in the document]
Comment (minalui) [reply src/app/error.tsx:1]: OK!

File: src/app/clients/ClientsPage.tsx [hunk 3]
```diff
@@ -162,9 +160,10 @@ const ClientsPage: React.FC<{}> = () => {
             });
 
         return () => {
-            supabase.removeChannel(subscriptionChannel);
+            void supabase.removeChannel(subscriptionChannel);
         };
-    }, [startPoint, endPoint, primaryFilters, sortState]);
+    }, [startPoint, endPoint, primaryFilters, sortState, fetchAndDisplayClientsData]);
```
Comment (rickaw) [inline src/app/clients/ClientsPage.tsx:165]: I don't think you need this change anymore since we don't call fetchAndDisplayClientsData?

**Review by rickaw — APPROVED** [review]
one last thing but feel free to merge once that's addressed!

File: src/app/clients/ClientsPage.tsx (hunk 3 above)
Comment (rickaw) [reply src/app/clients/ClientsPage.tsx:165]: ah that makes sense!

**Pull request merged**